pkgs/top-level/cuda-packages.nix

```diff
@@ -123,4 +123,6 @@ let
     fixedPoints.extends composedExtension passthruFunction
   );
 in
-cudaPackages
+lib.warnIf (lib.versionOlder cudaVersion "12.0" && config.allowAliases)
```

Comment:
```suggestion
lib.warnIf (lib.versionOlder cudaVersion "11.4" && config.allowAliases)
```

is safe

Reply:
In the PR/commit message I gave reasons for choosing 12.0 as a baseline rather than 11.4. By the time of 25.05’s release, 11.x will all depend on GCC versions that are unsupported upstream, which could result in old GCC versions sticking around solely for obsolete versions of CUDA.

Picking the policy listed in the release notes currently in this PR also helps us establish expectations for whether we will be obligated in future to keep an indefinitely‐growing list of versions of CUDA that haven’t been updated in years and only work with compilers that haven’t been updated in years and don’t have upstream support.

I think it’s better to set those expectations now, unless there’s a strong reason not to do it like this. My survey of the affected packages suggested that there’s only one package in Nixpkgs that is currently pinned to CUDA 11.x and that *may* not support 12.x (but I don’t know whether it does or not).

Reply:
Of course we should probably get TensorFlow off the unnecessary 11.x pin before actually merging this to avoid annoying users. I’m not sure I’m the best person for that job, though.

Reply:
I see. I must say I couldn't find any explicit information about gcc11's projected end of life, but I agree we don't need to support it any longer than upstream does.

Regarding updates for cuda11, maybe it doesn't needs any: the stack is somewhat rigid and stable, slow to evolve, the newer releases tend to just add support for new devices/features, and most applications (hpc) aren't security-critical...

Regarding users of cuda11, I'm afraid we (I) were rather hasty about discarding e.g. 1<=torch<2 which is used by a substantial amount of research code, and I don't think one could easily build it against cuda12.

Regarding tensorflow... nobody wants to deal with bazel

Reply:
In either case, thanks for picking this

Reply:
> I see. I must say I couldn't find any explicit information about gcc11's projected end of life, but I agree we don't need to support it any longer than upstream does.

Yes, unfortunately I couldn’t find a canonical place where the policy is written down. But looking at the pages for each version (e.g. [GCC 10](https://gcc.gnu.org/gcc-10/)) and poking at the Wayback Machine, the pattern is that the last minor release comes out after ~3 years, around the time of a new major release, and it’s marked as unsupported at the same time.

> Regarding users of cuda11, I'm afraid we (I) were rather hasty about discarding e.g. 1<=torch<2 which is used by a substantial amount of research code, and I don't think one could easily build it against cuda12.

I see; so this is about out‐of‐tree users that are manually repackaging unsupported versions of PyTorch? Since, as you said, the stack is quite stable and security is not of paramount concern, would these users be decently served by simply continuing to use old Nixpkgs versions that retain CUDA 11.x until they no longer need those versions of PyTorch? There’s still most of a year to go before the first version that would drop it will be released, and if updates aren’t necessary then the Nixpkgs versions going out of support shouldn’t be a huge deal.

> Regarding tensorflow... nobody wants to deal with bazel

Sounds like a challenge :) I’d hope that it’s as simple as just giving it a higher version, if the current TensorFlow releases are supposed to support them anyway, but build systems always contain wonderful new surprises. If me fumbling around trying to get that to work would help with this PR I’m happy to give it a shot.

Reply:
> would these users be decently served by simply continuing to use old Nixpkgs versions

That's the baseline and the answer mostly is yes. "Mostly" because what, in fact, hasn't been "slow and stable" is nixpkgs' `cudaPackages` which has gone through several refactorings and still occasionally suffers from e.g. various dynamic loading issues

> Sounds like a challenge :) I’d hope that it’s as simple as just giving it a higher version,

It's slightly more complicated than that and my impression is that several people have already burned out on this one (e.g. I don't suppose @zeuner feels particularly enthusiastic about this exercise)

Reply:
https://github.com/NixOS/nixpkgs/pull/334996 looks good to go for `python3Packages.tensorflow-bin`, at least. I guess updating the from‐source build is more complicated?

Reply:
> I’d hope that it’s as simple as just giving it a higher version

With a compatible cuDNN version in place (8.9?), it's probably worth a try.

Reply:
Okay, I’ll try giving it a shot soon. One thing confuses me: we list cuDNN 8.9 as supporting CUDA 12.2 at most, but TensorFlow pairs it with CUDA 12.3. What’s going on there? (And are those versions minimum requirements or precise pinned versions? TensorFlow only pins the major version right now.)

Reply:
Consider using our handy `cudaPackages.cudaOlder` instead of the calls to `lib` :)

Reply:
Let's proof read I guess.

>  we list cuDNN 8.9 as supporting CUDA 12.2 at most, 

Seems to check out:

https://docs.nvidia.com/deeplearning/cudnn/archives/cudnn-897/support-matrix/index.html

Reply:
![image](https://github.com/user-attachments/assets/65547846-eac0-43b1-bca2-581160aa14ca)

https://www.tensorflow.org/install/source#gpu

...well, yes